src/server/session.ts

```diff
@@ -304,6 +286,89 @@ namespace ts.server {
+    ): T[] {
+        const outputs: T[] = [];
+        combineProjectOutputWorker(projects, <sourcemaps.SourceMappableLocation><any>undefined, projectService, ({ project }, tryAddToTodo) => {
+            for (const output of action(project)) {
+                if (!contains(outputs, output, resultsEqual) && !tryAddToTodo(project, getLocation(output))) {
+                    outputs.push(output);
+                }
+            }
+        });
+        return outputs;
+    }
+
+    function combineProjectOutputForReferences(projects: Projects, initialLocation: sourcemaps.SourceMappableLocation, projectService: ProjectService): ReadonlyArray<ReferencedSymbol> {
+        const outputs: ReferencedSymbol[] = [];
+
+        combineProjectOutputWorker(projects, initialLocation, projectService, ({ project, location }, tryAddToTodo) => {
+            for (const outputReferencedSymbol of project.getLanguageService().findReferences(location.fileName, location.position) || emptyArray) {
+                let symbolToAddTo = find(outputs, o => documentSpansEqual(o.definition, outputReferencedSymbol.definition));
+                if (!symbolToAddTo) {
+                    symbolToAddTo = { definition: outputReferencedSymbol.definition, references: [] };
+                    outputs.push(symbolToAddTo);
+                }
+
+                for (const ref of outputReferencedSymbol.references) {
+                    if (!contains(symbolToAddTo.references, ref, documentSpansEqual) && !tryAddToTodo(project, documentSpanLocation(ref))) {
+                        symbolToAddTo.references.push(ref);
+                    }
+                }
+            }
+        });
+
+        return outputs.filter(o => o.references.length !== 0);
+    }
+
+    interface ProjectAndLocation {
+        readonly project: Project;
+        readonly location: sourcemaps.SourceMappableLocation;
+    }
+
+    function combineProjectOutputWorker(
+        projects: Projects,
+        initialLocation: sourcemaps.SourceMappableLocation,
+        projectService: ProjectService,
+        cb: (where: ProjectAndLocation, getMappedLocation: (project: Project, location: sourcemaps.SourceMappableLocation) => boolean) => void,
+    ): void {
+        const seenProjects = createMap<true>();
+        const filesToClose: string[] = [];
+        const toDo: ProjectAndLocation[] = (isProjectsArray(projects) ? projects : projects.projects).map(project => ({ project, location: initialLocation }));
+
+        while (toDo.length) {
+            const { project, location } = Debug.assertDefined(toDo.pop());
+            if (!addToSeen(seenProjects, project.projectName)) continue;
+            cb({ project, location }, (project, location) => {
+                const mapsTo = project.getSourceMapper().tryGetMappedLocation(location);
+                if (!mapsTo) return false;
+                const info = mapsTo && projectService.tryOpeningProjectForFileIfNecessary(toNormalizedPath(mapsTo.fileName));

```

Comment:
I think instead of opening a file, better approach would be to get project if exists otherwise get configFileName and create the project (you dont want inferred project in this scenario) so you need to handle case where project created by configFileName doesnt include the project? You would want to rather use addExternalProjectReference (rename it correctly) instead. We shouldnt open the script info since that has many implications with project Root etc. In this scenario what should be projectRoot? Since we dont want the search to go on to say "/user/username/" or some such?